File: workers/func2-ts/src/telemetry.ts
import { context, metrics, propagation, trace, type Attributes } from "@opentelemetry/api";
import { W3CTraceContextPropagator } from "@opentelemetry/core";
import { OTLPMetricExporter } from "@opentelemetry/exporter-metrics-otlp-http";
import { resourceFromAttributes } from "@opentelemetry/resources";
import { MeterProvider, PeriodicExportingMetricReader } from "@opentelemetry/sdk-metrics";
import { NodeTracerProvider } from "@opentelemetry/sdk-trace-node";

export interface TraceFields {
  span_id?: string;
  trace_flags?: string;
  trace_id?: string;
}

export function setupTelemetry(serviceName: string, otlpEndpoint: string) {
  const resource = resourceFromAttributes({ "service.name": serviceName });
  const exporter = new OTLPMetricExporter({ url: otlpEndpoint });
  const reader = new PeriodicExportingMetricReader({
    exporter,
    exportIntervalMillis: 5000,
  });
  const meterProvider = new MeterProvider({
    resource,
    readers: [reader],
  });

  metrics.setGlobalMeterProvider(meterProvider);

  const tracerProvider = new NodeTracerProvider({ resource });
  tracerProvider.register({
    propagator: new W3CTraceContextPropagator(),
  });

  return {
    meter: metrics.getMeter(serviceName),
    tracer: trace.getTracer(serviceName),
  };
}

export function extractParentContext(traceparent?: unknown) {
  if (typeof traceparent !== "string" || traceparent.trim() === "") {
    return context.active();
  }

  return propagation.extract(context.active(), { traceparent });
}

export function currentTraceFields(): TraceFields {
  const span = trace.getSpan(context.active());
  const spanContext = span?.spanContext();

  if (!spanContext) {
    return {};
  }

  return {
    span_id: spanContext.spanId,
    trace_flags: spanContext.traceFlags.toString(16).padStart(2, "0"),
    trace_id: spanContext.traceId,
  };
}

export function currentTraceparent(): string | undefined {
  const carrier: Record<string, string> = {};
  propagation.inject(context.active(), carrier);
  return carrier.traceparent;
}

export async function withTaskSpan<T>(
  tracer: ReturnType<typeof trace.getTracer>,
  name: string,
  traceparent: unknown,
  attributes: Attributes,
  fn: () => Promise<T>,
): Promise<T> {
  const parentContext = extractParentContext(traceparent);

  return context.with(parentContext, () =>
    tracer.startActiveSpan(name, { attributes }, async (span) => {
      try {
        return await fn();
      } catch (error) {
        span.recordException(error as Error);
        throw error;
      } finally {
        span.end();
      }
    }),
  );
}

File: workers/func1-python/src/telemetry.py
from __future__ import annotations

from opentelemetry import metrics, trace
from opentelemetry.exporter.otlp.proto.http.metric_exporter import OTLPMetricExporter
from opentelemetry.sdk.metrics import MeterProvider
from opentelemetry.sdk.metrics.export import PeriodicExportingMetricReader
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.trace.propagation.tracecontext import TraceContextTextMapPropagator


def setup_telemetry(service_name: str, otlp_endpoint: str) -> tuple:
    resource = Resource.create({"service.name": service_name})

    metric_exporter = OTLPMetricExporter(endpoint=otlp_endpoint)
    metric_reader = PeriodicExportingMetricReader(metric_exporter, export_interval_millis=5000)
    meter_provider = MeterProvider(resource=resource, metric_readers=[metric_reader])
    metrics.set_meter_provider(meter_provider)

    tracer_provider = TracerProvider(resource=resource)
    trace.set_tracer_provider(tracer_provider)

    return (
        metrics.get_meter(service_name),
        trace.get_tracer(service_name),
        TraceContextTextMapPropagator(),
    )
